kornia/enhance/equalization.py

```diff
@@ -0,0 +1,306 @@
+
+    # equalize tiles
+    flatten_interp_tiles: torch.Tensor = (interp_tiles * 255).long().flatten(-2, -1)  # B x GH x GW x 4 x C x (THxTW)
+    flatten_interp_tiles = flatten_interp_tiles.unsqueeze(-3).expand(num_imgs, gh, gw, 4, c, th * tw)
+    preinterp_tiles_equalized = torch.gather(
+        mapped_luts, 5, flatten_interp_tiles).reshape(num_imgs, gh, gw, 4, c, th, tw)  # B x GH x GW x 4 x C x TH x TW
+
+    # interp tiles
+    tiles_equalized: torch.Tensor = torch.zeros_like(interp_tiles, dtype=torch.long)
+
+    # compute the interpolation weights (shapes are 2 x TH x TW because they must be applied to 2 interp tiles)
+    ih = torch.arange(2 * th - 1, -1, -1, device=interp_tiles.device).div(2 * th - 1)[None].T.expand(2 * th, tw)
+    ih = ih.unfold(0, th, th).unfold(1, tw, tw)  # 2 x 1 x TH x TW
+    iw = torch.arange(2 * tw - 1, -1, -1, device=interp_tiles.device).div(2 * tw - 1).expand(th, 2 * tw)
+    iw = iw.unfold(0, th, th).unfold(1, tw, tw)  # 1 x 2 x TH x TW
+
+    # compute row and column interpolation weigths
+    tiw = iw.expand((gw - 2) // 2, 2, th, tw).reshape(gw - 2, 1, th, tw).unsqueeze(0)  # 1 x GW-2 x 1 x TH x TW
+    tih = ih.repeat((gh - 2) // 2, 1, 1, 1).unsqueeze(1)  # GH-2 x 1 x 1 x TH x TW
+
+    # internal regions
+    tl, tr, bl, br = preinterp_tiles_equalized[:, 1:-1, 1:-1].unbind(3)
+    t = tiw * (tl - tr) + tr
+    b = tiw * (bl - br) + br
+    tiles_equalized[:, 1:-1, 1:-1] = tih * (t - b) + b
+
+    # corner regions
+    tiles_equalized[:, 0::gh - 1, 0::gw - 1] = preinterp_tiles_equalized[:, 0::gh - 1, 0::gw - 1, 0]
+
+    # border region (h)
+    t, b, _, _ = preinterp_tiles_equalized[:, 1:-1, 0].unbind(2)
+    tiles_equalized[:, 1:-1, 0] = tih.squeeze(1) * (t - b) + b
+    t, b, _, _ = preinterp_tiles_equalized[:, 1:-1, gh - 1].unbind(2)
+    tiles_equalized[:, 1:-1, gh - 1] = tih.squeeze(1) * (t - b) + b
+
+    # border region (w)
+    l, r, _, _ = preinterp_tiles_equalized[:, 0, 1:-1].unbind(2)
+    tiles_equalized[:, 0, 1:-1] = tiw * (l - r) + r
+    l, r, _, _ = preinterp_tiles_equalized[:, gw - 1, 1:-1].unbind(2)
+    tiles_equalized[:, gw - 1, 1:-1] = tiw * (l - r) + r
+
+    # same type as the input
+    return tiles_equalized.to(interp_tiles).div(255.)
+
+
+def equalize_clahe(input: torch.Tensor, clip_limit: float = 40., grid_size: Tuple[int, int] = (8, 8)) -> torch.Tensor:
+    r"""Apply clahe equalization on the input tensor.
+
+    Args:
+        input (torch.Tensor): images tensor to equalize with shapes like :math:`(C, H, W)` or :math:`(B, C, H, W)`.
+        clip_limit (float): threshold value for contrast limiting. If 0 clipping is disabled. Default: 40.
+        grid_size (Tuple[int, int]): number of tiles to be cropped in each direction (GH, GW).
+
+    Returns:
+        torch.Tensor: Equalized image or images with shape as the input.
+
+    """
+    if not torch.is_tensor(input):
+        raise TypeError(f"Input input type is not a torch.Tensor. Got {type(input)}")
```

Comment:
add end line in between

Reply:
f"Input input type is not a torch.Tensor.\n Got {type(input)}"?

Reply:
Between if blocks